[FIX] website: convert bootstrap modal option

Goaman opened the pull request:
Before this commit, the popup snippets were configuring the modal option
for bootstrap 4 rather than 5.
The `data-focus` in bootstrap 4 becomes `data-bs-focus` in bootstap 5.

Without `data-bs-focus` option, the Newsletter popup snippet became
uneditable because any click inside popup changed the selection to be
inside the first focusable element within the modal.

The reason of the input being focused is: upon click inside the modal,
the target of the focusin event is the parent of the modal because that
parent is contenteditable="true". Because that target is outside the
modal, `_handleFocusin` correct the selection by focusing the text
input.

task-3102155




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/108091).

dmo-odoo commented:
@qsm-odoo This originated as an editor bug but it turns out it is the snippet template that has not been properly updated to BS5. However, this is 16.0. How do you generally fix this kind of stuff in stable versions ? Is there some migration script or something ? I mean, what about people who have dropped this snippet and have it in their website in its BS4 version ? It doesn't work anymore since BS5 ?

qsm-odoo commented:
@dmo-odoo :wave: 

- I did not know about (or forgot about it) this `data-focus` stuff
- I don't understand why we ever wanted `data-focus="false"` (is not it cool to have the focus on the modal when it opens? :thinking:). If no apparent justification while blaming & cie, maybe it could be nice to just remove it actually?
- The commit message does not speak about the task's original purpose so I wonder how this is related: of course `data-bs-focus` seems right but I don't have any problem with the newsletter popup as mentioned in the related task :grimacing: Could you enlighten me? :+1: 

As for stable fixes of that kind and migration, it's case by case. In this case:

- XML changes are not supposed to be merged in stable but in this case, I'd say it's ok. 16.0 is still young, I think we can take the risk of someone having made an xpath on this wrong attribute, especially since it would only break entering edit mode and not the rendering of the website. If you wanted to be extra safe though (as we personally try to do most of the time now), you could patch this in JS only. There are multiples:
    - Patch the `_computeSnippetTemplates` method (I recently introduced a `_patchForComputeSnippetTemplates` method to ease patching `website` stuff) to fix the received template before it is processed.
    - Patch the `onBuilt` option' method of modal to fix the data-focus into data-bs-focus before it is added in the DOM
    - Any other crappy patch that would fix stable, keep it stable, as long as fixed the proper way in master.

   (of course, if you patch in JS, I would add a comment on top of the template saying it is patched in JS)
   **I would however recommend patching in JS even if you choose to patch the XML directly** otherwise existing 16.0 won't receive the fix until they do a -u (which most won't do).

- No one has been migrated to 16.0 yet, so users with s_popup elements in previous versions are still fixable by patching the bootstrap migration script in the upgrade repo. You can probably add that in an upgrade PR linked directly to this one :pray: (tell me if help needed).

- For 16.0 users that already dropped the buggy s_popup snippet as it is right now:
   1. Is it really buggy without `data-bs-focus="false"`. If yes, should not that be fixed? (in the end, popup should be able to work with both, right? :thinking:)
   2. Depending on the bug, it's often acceptable to ask the customer to remove its snippet and drop a new one once the fix lands in 16.0. If not, then I guess any crappy JS fix at page rendering to fix the modal would do (if later removed in master with a migration script for 16.1 fixing users modals)... but this is rarely necessary.

Goaman commented:
@qsm-odoo 

> I don't understand why we ever wanted data-focus="false" (is not it cool to have the focus on the modal when it opens? 🤔). If no apparent justification while blaming & cie, maybe it could be nice to just remove it actually?

Removing the attribute does not make the first input to be focus when the modal opens. It does not either in the demo page of bootstrap Modal. Reading the source code, it should focus when focusin changes, which does not seems to be triggered when opening the modal nor clicking inside it in "readonly" (page not in edit mode).

> The commit message does not speak about the task's original purpose so I wonder how this is related: of course data-bs-focus seems right but I don't have any problem with the newsletter popup as mentioned in the related task 😬 Could you enlighten me? 👍

I've updated the commit message: 
> Without data-bs-focus option, the Newsletter popup snippet became
uneditable because any click inside popup changed the selection to be
inside the first focusable element within the modal.

The reason is that the focusin event is triggered on a parent of the modal that has the attribute `contenteditable="true"`.
As the code of the modal focus trap specifies (see `_handleFocusin`), if the target is a parent of the modal, it should "trap" the event and focus the first focusable element within the modal (in our case, the input text).

I've fixed the case in JS with the `onBuilt` method as you suggested. I don't think we need to make an upgrade PR as the problems only occurs when trying to edit popup snippets. `onBuilt` is enough. There is not problem in "readonly".

qsm-odoo commented:
> I don't think we need to make an upgrade PR as the problems only occurs when [...]

So it is required if a bug can occur :grimacing: We can do it if you prefer but I really think it's a matter of a few minutes checking the existing BS5 migration script and attach one upgrade PR to this one... :wink: 

As for the `onBuilt` solution, are you sure you need to do it both for popup generically + newsletter popup specifically? (are not newsletter popups fixed by the generic popup solution?)

dmo-odoo commented:
> > I don't think we need to make an upgrade PR as the problems only occurs when [...]
> 
> So it is required if a bug can occur 😬 We can do it if you prefer but I really think it's a matter of a few minutes checking the existing BS5 migration script and attach one upgrade PR to this one... 😉

I think what @Goaman meant is that since the problem can only occur in edit mode, you'll have the `onBuilt` fix anyway. There is no combination of factors that can simultaneaously trigger the bug _and_ **not** have the `onBuilt` fix. So in a way it is correct to say that the `onBuilt` fix is sufficient since the bug can not happen anymore, regardless of if you dropped the snippet before or not. However, I agree that adding the fix to the BS5 upgrade script is still a good idea. Not because the problem can still occur (since I think it cannot) but because if we don't add it to the upgrade script then we will never be confident that we can remove the js fix, and it would therefore linger forever in the code, which is obviously not something we want.

> As for the `onBuilt` solution, are you sure you need to do it both for popup generically + newsletter popup specifically? (are not newsletter popups fixed by the generic popup solution?)

Unfortunately no, those two snippet seem to be completely unrelated classes.

qsm-odoo commented:
@dmo-odoo 

> There is no combination of factors that can simultaneaously trigger the bug and not have the onBuilt fix

`onBuilt` is only called when a new snippet is added in the page so if you click on a non-properly migrated snippet, you will still have the bug IMO. Maybe you could re-test to double-check? :wink: But yes, in any case, we would still need the upgrade script to have the proper DOM / and be able to remove any JS fix in master as you explained.

> Unfortunately no, those two snippets seem to be completely unrelated classes.

I just tested... the JS code is website_mass_mailing is not needed... The newsletter popup being a popup, the first `onBuilt` (generic fix for all popups) is enough as I suspected :grimacing:

dmo-odoo commented:
> > Unfortunately no, those two snippets seem to be completely unrelated classes.
> 
> I just tested... the JS code is website_mass_mailing is not needed... The newsletter popup being a popup, the first `onBuilt` (generic fix for all popups) is enough as I suspected 😬

I'm not sure I understand how this works then. The first `onBuilt` override is in `options.registry.mailing_list_subscribe` which extends `options.Class` and the other is in `options.registry.mailing_list_subscribe` which also extends `options.Class`. I mean they are two unrelated classes, how comes the code of one affects the other ? I'm very confused.

qsm-odoo commented:
@dmo-odoo Specific options target specific DOM (data-js + data-selector system). You have one option targeting all popups (`data-js="SnippetPopup" data-selector=".s_popup"`) and one other unrelated option targeting newletter popups (`data-js="mailing_list_subscribe" data-selector=".o_newsletter_popup"`). Both options are thus targeting newsletter popups since newsletter popups are popups. Basically, you could say it is a mix of behaviors by composition, not by inheritance.

dmo-odoo commented:
@robodoo r+
